handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -2384,11 +2376,18 @@ private SSLSessionBindingEvent newSSLSessionBindingEvent(String name) {
         }
 
         @Override
-        public void setSessionId(OpenSslSessionId sessionId) {
+        public void setSessionDetails(
+                long creationTime, long lastUpdated, OpenSslSessionId sessionId,
+                Map<String, Object> keyValueStorage) {
             synchronized (ReferenceCountedOpenSslEngine.this) {
                 if (this.id == OpenSslSessionId.NULL_ID) {
                     this.id = sessionId;
-                    creationTime = System.currentTimeMillis();
+                    this.creationTime = creationTime;
+                    lastAccessed = lastUpdated;
+                    // Update the key value storage. It's fine to just drop the previous stored values on the floor
+                    // as the JDK does the same in the sense that it will use a new SSLSessionImpl instance once the
+                    // handshake was done
+                    this.keyValueStorage = keyValueStorage;
```

Comment:
I'll need some time to think about this line...

Reply:
Sure ... we can remove this as well it's just consistent with the JDK it seems

Reply:
Yeah, this is problematic, because this happens _after_ the `TrustManager` authenticates the peer, and the trust manager may access the handshake session and put stuff there, which this line will overwrite. We need to clear the session storage before - or as part of - the `AbstractCertificateVerifier.verify()` call. And then we need to make sure the values added in the handshake session are preserved in the native session, which I guess would happen in here.

I made a patch to `SSLEngineTest` to capture this scenario – this patch passes for `JdkSslEngineTest` but fails for various OpenSsl engine tests. Not sure if you wanted to fix that in this PR (given the PR title) or in a separate PR. Here's the test patch:

```diff
Index: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java b/handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java
--- a/handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java	(revision 12cd670d7a8e471b0ea4911469daeb57723bd85b)
+++ b/handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java	(date 1707066361024)
@@ -3621,8 +3621,57 @@
                 clientContextBuilder.keyManager(ssc.key(), ssc.cert());
             }
         }
+        final String handshakeKey = "handshake";
+        TrustManagerFactory tmf = new SimpleTrustManagerFactory() {
+            @Override
+            protected void engineInit(KeyStore keyStore) throws Exception {
+            }
+
+            @Override
+            protected void engineInit(ManagerFactoryParameters managerFactoryParameters) throws Exception {
+            }
+
+            @Override
+            protected TrustManager[] engineGetTrustManagers() {
+                return new TrustManager[] {
+                        new X509ExtendedTrustManager() {
+                            @Override
+                            public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+                            }
+
+                            @Override
+                            public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+                            }
+
+                            @Override
+                            public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+                                engine.getHandshakeSession().putValue(handshakeKey, Boolean.TRUE);
+                            }
+
+                            @Override
+                            public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+                                engine.getHandshakeSession().putValue(handshakeKey, Boolean.TRUE);
+                            }
+
+                            @Override
+                            public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
+                            }
+
+                            @Override
+                            public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
+                            }
+
+                            @Override
+                            public java.security.cert.X509Certificate[] getAcceptedIssuers() {
+                                return new java.security.cert.X509Certificate[0];
+                            }
+                        }
+                };
+            }
+        };
+
         clientSslCtx = wrapContext(param, clientContextBuilder
-                                        .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                                        .trustManager(tmf)
                                         .sslProvider(sslClientProvider())
                                         .sslContextProvider(clientSslContextProvider())
                                         .protocols(param.protocols())
@@ -3635,7 +3684,7 @@
         if (mutualAuth) {
             serverContextBuilder.clientAuth(ClientAuth.REQUIRE);
         }
-        serverSslCtx = wrapContext(param, serverContextBuilder.trustManager(InsecureTrustManagerFactory.INSTANCE)
+        serverSslCtx = wrapContext(param, serverContextBuilder.trustManager(tmf)
                                      .sslProvider(sslServerProvider())
                                      .sslContextProvider(serverSslContextProvider())
                                      .protocols(param.protocols())
@@ -3710,6 +3759,17 @@
 
             Object value = new Object();
 
+            assertEquals(1, clientSession.getValueNames().length);
+            assertEquals(clientSession.getValue(handshakeKey), Boolean.TRUE);
+            clientSession.removeValue(handshakeKey);
+
+            if (mutualAuth) {
+                // Server trust manager factory is only called if server authenticates clients.
+                assertEquals(1, serverSession.getValueNames().length);
+                assertEquals(serverSession.getValue(handshakeKey), Boolean.TRUE);
+                serverSession.removeValue(handshakeKey);
+            }
+
             assertEquals(0, clientSession.getValueNames().length);
             clientSession.putValue("test", value);
             assertEquals("test", clientSession.getValueNames()[0]);
```

Reply:
Yeah, this is not going to fly. Let's talk offline.